pisa-proxy/proxy/strategy/src/discovery/discovery.rs

```diff
@@ -0,0 +1,124 @@
+    },
+    readwritesplitting::ReadWriteEndpoint,
+};
+
+//define discovery kind (support MHA,RDS,MGR etc.)
+pub enum DiscoveryKind {
+    MasterHighAvailability(DiscoveryMasterHighAvailability),
+}
+
+pub trait Discovery {
+    type Output;
+
+    fn build(config: MasterHighAvailability, rw_endpoint: ReadWriteEndpoint) -> Self::Output;
+    fn build_monitors(
+        &self,
+        monitor_response_channel: crate::readwritesplitting::MonitorResponseChannel,
+    ) -> Vec<MonitorKind>;
+}
+
+pub struct DiscoveryMasterHighAvailability {
+    config: MasterHighAvailability,
+    rw_endpoint: ReadWriteEndpoint,
+    pub monitors: Vec<MonitorKind>,
+}
+
+impl Discovery for DiscoveryMasterHighAvailability {
+    type Output = Self;
+
+    fn build(config: MasterHighAvailability, rw_endpoint: ReadWriteEndpoint) -> Self::Output {
+        Self { config, rw_endpoint, monitors: vec![] }
+    }
+
+    fn build_monitors(
+        &self,
+        monitor_response_channel: crate::readwritesplitting::MonitorResponseChannel,
+    ) -> Vec<MonitorKind> {
+        let mut monitors = vec![];
+        monitors.push(MonitorKind::Connect(MonitorConnect::new(
+            self.config.user.clone(),
+            self.config.password.clone(),
+            self.config.connect_period,
+            self.config.connect_timeout,
+            self.config.connect_failure_threshold,
+            self.rw_endpoint.clone(),
+            monitor_response_channel.monitor_response_tx.clone(),
+        )));
+        monitors.push(MonitorKind::Ping(MonitorPing::new(
+            self.config.user.clone(),
+            self.config.password.clone(),
+            self.config.ping_period,
+            self.config.ping_timeout,
+            self.config.ping_failure_threshold,
+            monitor_response_channel.monitor_response_tx.clone(),
+            self.rw_endpoint.clone(),
+        )));
+        monitors.push(MonitorKind::ReadOnly(MonitorReadOnly::new(
+            self.config.user.clone(),
+            self.config.password.clone(),
+            self.config.read_only_period,
+            self.config.read_only_timeout,
+            self.config.read_only_failure_threshold,
+            monitor_response_channel.monitor_response_tx.clone(),
+            self.rw_endpoint.clone(),
+        )));
+        monitors.push(MonitorKind::Lag(MonitorReplicationLag::new(
+            self.config.user.clone(),
+            self.config.password.clone(),
+            self.config.replication_lag_period,
+            self.config.replication_lag_timeout,
+            self.config.replication_lag_failure_threshold,
+            self.config.max_replication_lag,
+            monitor_response_channel.monitor_response_tx,
+            self.rw_endpoint.clone(),
+        )));
+
+        monitors
+    }
+}
+
+#[derive(Debug)]
+pub enum MonitorKind {
+    Connect(MonitorConnect),
+    Ping(MonitorPing),
+    Lag(MonitorReplicationLag),
```

Comment:
Better keep it same with name ReplicationLag

Reply:
update to ReplicationLag